rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,775 @@
+    """Combines multiple dense or sparse feature tensors into one.
+
+    This layer combines features by following these steps:
+    1. Apply a `ConcatenateSparseDenseFeatures` layer separately to sequence- and
+        sentence-level features, yielding two tensors (one for each feature type).
+    2. Apply a `ConcatenateSequenceSentenceFeatures` layer to the two tensors to combine
+        sequence- and sentence-level features into one tensor.
+
+    Arguments:
+        attribute: Name of attribute (e.g. `text` or `label`) whose features will be
+            processed.
+        attribute_signature: A dictionary containing two lists of `FeatureSignature`s, 
+            one for each feature type of the given attribute.
+        config: A model config used for correctly parametrising the `ConcatenateSparse
+            DenseFeatures` and `ConcatenateSequenceSentenceFeatures` layers.
+
+    Input shape:
+        Tuple of four tensor inputs:
+            sequence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, max_seq_length, input_dim)` where `input_dim` can be
+                different for sparse vs dense tensors.
+            sentence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, 1, input_dim)` where `input_dim` can be different for
+                sparse vs dense tensors, and can differ from that in `sequence_features`.
+            mask_sequence: dense 3-D tensor with the shape `(batch_size, max_sequence_
+                length, 1)`.
+            mask_combined_sequence_sentence: dense 3-D tensor with the shape `(batch_size,
+                max_sequence_length+1, 1)`, i.e. with the 2nd dimension combining the
+                lengths of sequence- and sentence-level features.
+
+    Output shape:
+        3-D tensor with shape: `(batch_size, sequence_length, units)` where `units` is 
+        completely  determined by the internally applied `ConcatenateSparseDenseFeatures`
+        layer and `sequence_length` is completely determined by the internally applied 
+        `ConcatenateSequenceSentenceFeatures` layer.
+
+    Raises:
+        A ValueError if no feature signatures are provided.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        attribute_signature: Dict[Text, List[FeatureSignature]],
+        config: Dict[Text, Any],
+    ) -> None:
+        if not attribute_signature or not (
+            len(attribute_signature.get(SENTENCE, [])) > 0
+            or len(attribute_signature.get(SEQUENCE, [])) > 0
+        ):
+            raise ValueError("The data signature must contain some features.")
+
+        super().__init__(name=f"rasa_input_layer_{attribute}")
+
+        self._tf_layers = {}
+
+        # 1. prepare layers for combining sparse and dense features
+        for feature_type in attribute_signature:
+            if len(attribute_signature[feature_type]) == 0 or feature_type not in [
```

Comment:
when can `len(...) == 0`?

Reply:
I mean, when a sequence attribute like TEXT won't have any sentence-level features?

Reply:
do you mean `TEXT` would be empty? then it shouldn't exist

Reply:
No. OK, maybe I don't understand how our features work. Let's take a step back: Do we have or allow featurisers that create only sequence-level features?

Reply:
yes, it happens for labels un diet, but then why would `attribute_signature[feature_type]` be empty list?

Reply:
Well, I don't know, that was my assumption for when there are no features of the given feature type. I think you had a discussion with Daksh around this: Was the conclusion that there should be an empty list, or that the feature type key should not be in the dict at all?

Reply:
> feature type key should not be in the dict at all
this one

Reply:
did you `print` `attribute_signature[feature_type]`? is it ever empty?

Reply:
I did print out these things a long time ago but I didn't experiment with different combinations of sequence- and sentence-level features.